lib/services/messenger/datahubs/flowdock.ts

```diff
@@ -0,0 +1,124 @@
+import { Session } from 'flowdock';
+import * as _ from 'lodash';
+import { FlowdockConnectionDetails, FlowdockMessage } from '../../flowdock-types';
+import { DataHub, DataHubError, DataHubErrorCode } from './datahub';
+
+/**
+ * A service retrieves values from the flowdock 1-1 history
+ */
+export class FlowdockDataHub implements DataHub {
+	private session: Session;
+	private organization: string;
+
+	constructor(data: FlowdockConnectionDetails) {
+		this.session = new Session(data.token);
+		// The flowdock service both emits and calls back the error
+		// We'll just ignore the emit to prevent it bubbling
+		// This is not logged because a failure to find a value in this service could be deliberate
+		this.session.on('error', () => { /* do nothing */});
+		this.organization = data.organization;
+	}
+
+	/**
+	 * Search for the specified value set by a user.
+	 * @param user    Username to search associated with.
+	 * @param service The service for which we seek a value, eg discourse
+	 * @param key     Name of the value to retrieve.
+	 * @returns       Promise that resolves to the value.
+	 */
+	public fetchValue(user: string, service: string, key: string): Promise<string> {
+		// Retrieve a particular regex from the 1-1 message history of the user
+		const findKey = new RegExp(`My ${service} ${key} is (\\S+)`, 'i');
+		return this.fetchPrivateMessages(user, findKey)
+		.then((valueArray) => {
+			const value = valueArray[valueArray.length - 1].match(findKey);
+			if (value) {
+				return value[1];
+			}
+			throw new DataHubError(DataHubErrorCode.ValueNotFound, `Could not find value $key for $user`);
+		});
+	}
+
+	/**
+	 * Search for recent private messages with our account that match on username and regex.
+	 * @param username  Scope of the private messages to search.
+	 * @param filter    Narrow our search to just matches.
+	 * @returns         Promise that resolves to the message strings.
+	 */
+	private fetchPrivateMessages(username: string, filter: RegExp): Promise<string[]> {
+		// Fetch the id then 1-1 history associated with the username
+		return this.fetchUserId(username)
+		.then((userId) => {
+			const match = filter.source.match(/^([\w\s]+)/i);
+			return this.fetchFromSession(`/private/${userId}/messages`, match ? match[1] : undefined)
+			.then((fetchedMessages) => {
+				// Prune and clean the message history to text of interest
+				return _.filter(fetchedMessages, (message: FlowdockMessage) => {
+					return filter.test(message.content);
+				}).map((message: FlowdockMessage) => {
+					return message.content;
+				});
+			});
+		});
+	}
+
+	/**
+	 * Fetch a user's id from their username.
+	 * @param username  Username to search for.
+	 * @returns         id of the user.
+	 */
+	private fetchUserId = (username: string): Promise<string | undefined> => {
+		// Get all the users of the service
+		return this.fetchFromSession(`/organizations/${this.organization}/users`)
+			.then((foundUsers) => {
```

Comment:
Some of the promise clauses have indented `then`s and some don't. Don't mind either way, but a consistent style needs to be used.